### ISSUE TITLE
Option: Selectively clean up Polly Context?

### ISSUE THREAD

**Issue by reisenberger** [aspnet/HttpClientFactory#94]
Revisiting [this conversation](https://github.com/aspnet/HttpClientFactory/pull/65#discussion_r170357438) (click also "Show outdated") about cleaning up Polly execution `Context`.  I don't think we necessarily have to change anything - the below doesn't affect any recommended HttpClientFactory usage pattern afaics - but I'm airing an option so we've had the discussion.

---

Where we currently [do this](https://github.com/aspnet/HttpClientFactory/blob/60a23380a354ebaf27c8cd3505e745bd8b7f1541/src/Microsoft.Extensions.Http.Polly/PolicyHttpMessageHandler.cs#L117-L125), we could:

```csharp
var cleanUpContext = false;
var context = request.GetPolicyExecutionContext();
if (context == null)
{
    context = new Context();
    request.SetPolicyExecutionContext(context);
    cleanUpContext = true;
}

var policy = _policy ?? SelectPolicy(request);
HttpResponseMessage response = await policy.ExecuteAsync((c, ct) => SendCoreAsync(request, c, ct), context, cancellationToken);

if (cleanUpContext)
{
    request.SetPolicyExecutionContext(null);
}

return response;
```

We cannot clean up `Context` on exiting every `PolicyHttpMessageHandler`. We need to flow `Context` inwards and outwards through nested `PolicyHttpMessageHandler`s so that telemetry for an entire execution (including retries) correlates with the unique [`CorrelationId` on `Context`](https://github.com/App-vNext/Polly/wiki/Keys-and-Context-Data#correlationid).

But if we created the `Context`, we know both that it was not user-supplied, and that we are the outermost `PolicyHttpMessageHandler`, so we could clean it up.

**Philosophically** this feels slightly cleaner, in adhering to the principle of least side-effect: whether the overall call thru `HttpClient` was supplied an `HttpRequestMessage` with or without context, we return it in the same state. It's possible this extra cleanliness could prevent us being bitten later in some as-yet-unforeseen case.

**Practically**-speaking, as you hinted @rynowak, it feels like only some weird edge-cases could be affected.  An example could be where a user was re-using an `HttpRequestMessage` by deep cloning.  Cloning has been a solution to certain Use Cases in the past (eg [1](https://stackoverflow.com/questions/25044166/how-to-clone-a-httprequestmessage-when-the-original-request-has-content); [2](https://stackoverflow.com/questions/18000583/re-send-httprequestmessage-exception/18014515?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa)).  Tho some of those Use Cases will have been around orchestrating retries, for which of course we'll be recommending the `PolicyHttpMessageHandler` approach.  Maybe others can see other cases.

---

If you do want the extra cleanliness, I have a commit ~here~ which we could PR in.  

**EDIT**: Updated (and squashed) the commit for consistency of using `var`:  [`f332d50`](https://github.com/reisenberger/HttpClientFactory/commit/f332d500816cc59ca5c9b6979cbdf5b2461d7c2d)

**Comment by rynowak** [comment]
After thinking about this, it seems good to have this. If you get a chance to send a PR today or over the weekend I'll merge it.

If you won't get to this then I'll do it myself :)

Sunday is the last day for us to take ad-hoc changes without needing to ask for approval

**Comment by rynowak** [comment]
Fixed by 61d0983 - Thanks!
